block-storage/src/main/scala/coop/rchain/blockstorage/FileLMDBIndexBlockStore.scala

```diff
@@ -0,0 +1,255 @@
+        _                            <- Sync[F].delay { blockMessageRandomAccessFile.setLength(0) }
+        _ <- withWriteTxn { txn =>
+              index.drop(txn)
+            }
+      } yield ()
+    )
+
+  override def close(): F[Unit] =
+    lock.withPermit(
+      for {
+        blockMessageRandomAccessFile <- blockMessageRandomAccessFileRef.get
+        _                            <- Sync[F].delay { blockMessageRandomAccessFile.close() }
+      } yield ()
+    )
+}
+
+object FileLMDBIndexBlockStore {
+  private val checkpointPattern: Regex = "([0-9]+)-([0-9]+)".r
+
+  case class Config(
+      storagePath: Path,
+      indexPath: Path,
+      checkpointsDirPath: Path,
+      mapSize: Long,
+      maxDbs: Int = 1,
+      maxReaders: Int = 126,
+      noTls: Boolean = true
+  )
+
+  private[blockstorage] case class CheckpointIndex(
+      env: Env[ByteBuffer],
+      index: Dbi[ByteBuffer]
+  )
+
+  private[blockstorage] case class Checkpoint(
+      start: Long,
+      end: Long,
+      dirPath: Path,
+      storagePath: Path,
+      indexPath: Path,
+      index: Option[WeakReference[CheckpointIndex]]
+  )
+
+  private def loadCheckpoints[F[_]: Sync: Log](checkpointsDirPath: Path): F[List[Checkpoint]] =
+    for {
+      checkpointDirectories <- Sync[F].delay {
+                                checkpointsDirPath.toFile.mkdir()
+                                Files.list(checkpointsDirPath).filter(p => Files.isDirectory(p))
+                              }
+      checkpointDirectoriesList = checkpointDirectories
+        .collect(Collectors.toList[Path])
+        .asScala
+        .toList
+      checkpoints <- checkpointDirectoriesList.flatTraverse { dirPath =>
+                      dirPath.getFileName.toString match {
+                        case checkpointPattern(start, end) =>
+                          List(
+                            Checkpoint(
+                              start.toLong,
+                              end.toLong,
+                              dirPath,
+                              dirPath.resolve("storage"),
+                              dirPath.resolve("index"),
+                              None
+                            )
+                          ).pure[F]
+                        case other =>
+                          Log[F].warn(s"Ignoring directory '$other': not a valid checkpoint name") *>
+                            List.empty[Checkpoint].pure[F]
+                      }
+                    }
+      sortedCheckpoints = checkpoints.sortBy(_.start)
+      result <- if (sortedCheckpoints.headOption.forall(_.start == 0)) {
+                 if (sortedCheckpoints.isEmpty ||
+                     sortedCheckpoints.zip(sortedCheckpoints.tail).forall {
+                       case (current, next) => current.end == next.start
+                     }) {
+                   sortedCheckpoints.pure[F]
+                 } else {
+                   Sync[F].raiseError(
```

Comment:
We should not use `raiseError`, this is equivalent to throwing an exception. You can not potentially recover from it (unless you use `attempt` on the caller side). But this essentially bring us to imperative programming at this point - because requires using `attempt` on the caller side each time I'm using a function that returns `F[_]:Sync`. 

If you have a logical error that should not stop the system you should do one of the folowing:
1. return an error as a value (e.g `Either[Error, Value]`)
2. use `ApplicativeError` that is based on some ADT of errors (not exceptions that will be thrown by Task)

In other words: `Sync[F].raiseError` should **only** be used when the application should just simply crash. No recover, nothing. Just plain, old dead process. 
If that is what you were planning (stopping the system), potential risk that needs to be taken into consideration is resource cleanup - using for example `bracket` when calling `FileLMDBIndexBlockStore.create` . This is one more time a situation were we are not taking leverage of using FP paradigm - this is getting us closer to imperative hell. Why? Well, nothing is stopping the call to just use `FileLMDBIndexBlockStore.create` without taking care of resources.

Reply:
I have replaced most of the creation methods' return types with `Either[StorageError, BlockStore[F]]`  and got rid of `Sync[F].raiseError` calls.

Regarding resource cleanup: I can make `create` methods to return a `Resource[F, BlockStore[F]]`. Which would basically force the caller to cleanup the block store resources. WDYT?

Reply:
👍